### ISSUE TITLE
Allow users to export animated GIFs

### ISSUE THREAD

**Issue by fdb** [nodebox/seed#16]
When creating an animation, it would be nice if users could export the animation as an animated GIF.

This is challenging for a number of reasons:
- The output are HTML / SVG elements, not a canvas tag, so we can't use something like [GIF.js](https://jnordberg.github.io/gif.js/)
- The element we want to convert to GIF can be any element, of any size. The size can even change between frames.

One idea is to create a server-side renderer that could take seed sketches and render them in a headless browser, like [Headless Chrome](https://developers.google.com/web/updates/2017/04/headless-chrome) or [PhantomJS](http://phantomjs.org/). This would work as a webservice: we request a URL and it returns a GIF image.

**Comment by kunal-mohta** [comment]
Are you planning to render the animation, then screencast it and finally convert it into a GIF using some other webservice?

**Comment by fdb** [comment]
I would suggest creating a special URL route that *only* shows the output (like the embed view, but without the source editor and navigation bar), then use Chrome Headless to screencapture the page, then use something like GIF.js to generate a GIF file out of that.

This requires setting up a custom webservice (preferably using Heroku or Docker) that can convert sketches to GIF on-demand.

**Comment by kunal-mohta** [comment]
Okay, so basically making an API?

**Comment by fdb** [comment]
Yes, a basic webservice that takes in a Seed URL or Sketch address and outputs a GIF image. This could be the API address:

https://seedgif.herokuapp.com/create?sketch=-L4uLB99JzFHMPcVoiTm

**Comment by fdb** [comment]
Here's some documentation on running Headless Chrome on Heroku:
https://timleland.com/headless-chrome-on-heroku/

The official Heroku buildpack looks promising:
https://elements.heroku.com/buildpacks/minted/heroku-buildpack-chrome-headless

**Comment by kunal-mohta** [comment]
I have had a look over the links that you provided. 
It seems like the first focus should be on being able to make and run the app locally.

**Comment by kunal-mohta** [comment]
I was trying headless chrome locally using [Puppeteer](https://github.com/GoogleChrome/puppeteer)
but I am facing this error
![screenshot 2018-02-22 09 51 19](https://user-images.githubusercontent.com/30414954/36520343-803ee904-17b6-11e8-90f4-dec4a2cadaf5.png)

My code is this

![screenshot 2018-02-22 09 55 03](https://user-images.githubusercontent.com/30414954/36520359-a11b2426-17b6-11e8-8c93-c920ac52bb69.png)

Are you familiar with this kind of errors?
I am not being able to figure a way out of this.

**Comment by fdb** [comment]
No idea. Could you try a different site? Maybe google.com does not support headless browsing?

**Comment by kunal-mohta** [comment]
Let me try.

**Comment by kunal-mohta** [comment]
Nope, got the same error at  [https://seed.emrg.be/]( https://seed.emrg.be/)

**Comment by fdb** [comment]
Works for me:

![seed](https://user-images.githubusercontent.com/8477/36540832-1eee9c3c-17dc-11e8-9c7a-4c22eceaacff.png)

Here's the code:
```
const puppeteer = require('puppeteer');

(async() => {
    const browser = await puppeteer.launch()
    const page = await browser.newPage();
    await page.goto('https://seed.emrg.be/');
    await page.screenshot({path: 'seed.png'});
    await browser.close();
})();

```

**Comment by fdb** [comment]
The main issue now is that the actual sketch page loads the script async, so it shows the "loading..." indicator when screenshotting the page:

![seed](https://user-images.githubusercontent.com/8477/36541013-d2637558-17dc-11e8-9f5a-84cbbd3c4e49.png)

I've also linked issue #32 since we don't want to screenshot the entire UI, just the viewer portion.

**Comment by kunal-mohta** [comment]
Wait, how are you able to work with `async - await` syntax in Node.js without `try - catch` phrase. 
Doesn't it report `UnhandledPromiseRejectionWarning`?

**Comment by fdb** [comment]
I don't seem to have any issues with it. I'm using:

```
$ node -v
v9.4.0
```

**Comment by kunal-mohta** [comment]
Mine is `v8.9.0`. :worried:

**Comment by fdb** [comment]
You could try using `nvm` to see if it's the version that causing issues.

**Comment by kunal-mohta** [comment]
I updated to `v9.5.0`.
Still gives the same error.

**Comment by fdb** [comment]
The ECONNRESET is an error that comes back from the network stack. It looks like the requests are blocked on the network level somehow — perhaps overly aggressive filtering by some router / firewall? 

I think you should try with different sites.

**Comment by kunal-mohta** [comment]
I tried several sites with no hope.
Could you send me your `package.json` file please?

**Comment by fdb** [comment]
Here's mine: 

```
{
  "name": "headless",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "puppeteer": "1.1.0"
  }
}
```

And here's the `index.js`:

```
const puppeteer = require('puppeteer');

(async() => {
    const browser = await puppeteer.launch()
    const page = await browser.newPage();
    await page.goto('https://seed.emrg.be/sketch/-L4uLB99JzFHMPcVoiTm');
    await page.screenshot({path: 'seed.png'});
    await browser.close();
})();
```

**Comment by kunal-mohta** [comment]
Still no success :confused:

**Comment by fdb** [comment]
Does it work on a different site? E.g. the site of your school?

**Comment by kunal-mohta** [comment]
Nope, I tried that too. Didn't work.
I will have to try it on one of my peers' system.

**Comment by kunal-mohta** [comment]
@fdb I was finally able to get puppeteer running (though was not able to solve the above issue).
Now I am able to take screenshots of the browser.
However, have you decided upon the method to be used to record the screen, or is it yet to be figured out?

**Comment by fdb** [comment]
Since the built-in animations are entirely deterministic, you could calculate the duration * FPS to get the amount of frames. Then, loop through each frame and set the internal frame counter and render out each image in sequence. Then, make a GIF out of this image sequence.

Initially it would be good to have this working for just a static image import, which is useful by itself. 

Here's a crappy mockup of how that UI might look:

![seed-export-menu](https://user-images.githubusercontent.com/8477/36937952-2a81e374-1f1b-11e8-9c1a-d2b66039fb75.png)

**Comment by kunal-mohta** [comment]
@fdb Yes, the static image part can be achieved now.
However, I doubt if puppeteer is so accurate to keep up with the FPS of animations. 🤔

**Comment by fdb** [comment]
For the animation I would just render out an image sequence with the frame counter set to a specific value. So we don't render animations realtime; instead we export individual images at the exact frame time we want. We can then stitch these image sequences together using GIF.js or ffmpeg.

**Comment by kunal-mohta** [comment]
Okay, I understand your method.
But how can we account for the delay that will exist between the script calling the `screenshot` method and Puppeteer actually taking the screenshot?

**Comment by fdb** [comment]
None of it needs to be realtime. We can just pass the frame we want to export as a URL parameter, e.g. 
https://seed.emrg.be/view/SKETCH_ID_HERE?seed=ABC&frame=5  Then do this repeatedly for every frame.

**Comment by kunal-mohta** [comment]
Ohh! I get it now. 
Very nice! 👍
